Exclude Properties Using Exclusion List #113

deev-pal08 opened the pull request:
# Exclude Properties from URI Using Exclusion List

Exclusion List is a feature which allows you to exclude/restrict certain entities and their respective properties while generating URL payloads for entity_sets parsed by pyodata from the metadata.xml file

This feature has been implemented in the DirectBuider class of the odfuzz/entities.py file.

Earlier when we mentioned a particular property in the ExclusionList, it was only getting restricted from the Body of the generated Payload, but it should also be restricted from the URI of the generated payloads.

While using this Exclusion_List we can mention certain properties for a particular EntitySet which we do not want while generating the payloads.

**Example**:

 ```json
              {
                "$ENTITY_SET$":{
                        "Products": {
                            "Properties":["ProductID", "ProductName"],
                            "Nav_Properties":["Category"]
                        }
                    }
                }
```

Two functions use these properties to generate the payloads and include them in the URI:
1. **$filter**
2. **$orderby**

Implemented regex_match function in two of the classes which uses $orderby and $filter functions to generate the payloads for the URI:

1. **FilterQuery**: Class to generate queries for $filter function
2. **OrderbyQuery**: Class to generate queries for $orderby function

Both the classes randomly choose properties from the list of all the available properties of that particular entity set to generate the payloads.

After the properties are matched from the properties mentioned in the Exclusion List, they are removed from the List of properties used for generating queries for both the functions.

Also made some changes to the earlier implemented code for Exclusion List in the DirectBuilder class of the entities.py file in the regex_match and apply_restrictions methods for improving the Exclusion_List feature for EntitySets and Properties appearing in the Body.

deev-pal08 committed:
Exclude Properties Using Exclusion List #113

1. Implemented Exclusion List feature for properties appearing on the URI of the generated payloads for two different functions $filter and $orderby.

2. Used the regex_match function to match the properties with the ones mentioned in the ExclusionList and then remove it from the list of all the properties partricular to an EntitySet.

deev-pal08 committed:
Exclude Properties Using Exclusion List #113

1. Implemented Exclusion List feature for properties appearing on the URI of the generated payloads for two different functions $filter and $orderby.

2. Used the regex_match function to match the properties with the ones mentioned in the ExclusionList and then remove it from the list of all the properties partricular to an EntitySet.

deev-pal08 committed:
Update test_url_generator_only.py

Added new unit tests for testing properties restrictions feature where specific properties for specific entity_sets can be excluded while generating the URI from the $filter and $orderby query_option of entity_sets

Three new functions where included:

1. builder_with_restrictions(method_name): For generating query_strings and query_body with needed restrictions to entity_sets and properties.

2. test_direct_builder_filter_query_option(): For testing whether all the properties mentioned in the exclusion_list is been properly removed from the $filter query_options in the URI

3. test_direct_builder_orderby_query_option(): For testing whether all the properties mentioned in the exclusion_list is been properly removed from the $orderby query_options in the URI

deev-pal08 committed:
Update test_url_generator_only.py

Added new unit tests for testing properties restrictions feature where specific properties for specific entity_sets can be excluded while generating the URI from the $filter and $orderby query_option of entity_sets

Three new functions where included:

1. builder_with_restrictions(method_name): For generating query_strings and query_body with needed restrictions to entity_sets and properties.

2. test_direct_builder_filter_query_option(): For testing whether all the properties mentioned in the exclusion_list is been properly removed from the $filter query_options in the URI

3. test_direct_builder_orderby_query_option(): For testing whether all the properties mentioned in the exclusion_list is been properly removed from the $orderby query_options in the URI

deev-pal08 committed:
Merge branch 'Properties_Exclusion' of https://github.com/deev-pal08/odfuzz into Properties_Exclusion

deev-pal08 committed:
Updated Units tests in test_url_generator_only.py

Added units tests in test_url_generator_only.py file for testing the ExclusionList feature implemented to restrict various enitiy_sets and properties of those entity_sets from being added to the generated URI and Body.

Two Different methods have been added:

1. test_direct_builder_exclusion_list_Uri(): This method tests for the URI generated by ODFuzz for various methods both before and after passing the ExclusionList as the Restriction Object.

2. test_direct_builder_exclusion_list_body(): This method tests for the body generated by ODFuzz for various methods both before and after passing the ExclusionList as the Restriction Object.

deev-pal08 committed:
Updated tests cases #122

Refactored the unit tests in test_url_generator_only.py file to make the tests smaller and more readable.
Used parametrized tests, so same test structure use different inputs.
The @pytest.mark.parametrize will create essentially new test instance for every line of input list

phanak-sap commented on odfuzz/entities.py:
```diff
@@ -692,9 +690,18 @@ def get_depending_data(self):
     """The search query option."""
 
-    def __init__(self, entity, restrictions):
+    def __init__(self, entity, restrictions, restrictions_group):
```
Why is here new parameter, restriction_group, while in DirectBuilder the same data are provided just as substructure of restrictions? I would expected that nothing would change here as well.

deev-pal08 replied:
The first Parameter (restrictions) calls the forbidden_options method of the class RestrictionGroup which does the not provide the Exclusion_List we require to restrict the Properties. This is just a subpart of the RestrictionsGroup.

It provides us with the YAML file restriction Object:
Exclude:
    $filter:
        Categories:
            - Description
            - Picture
        Employees:
            - Photo
            - HireDate
        $F_ALL$:
            - replace
    $orderby:
        Categories:
            - Picture
            - Description
       
What we actually need is the restrictions passed as a json object to the RestrictionGroup Class when we initialize the RestrictionGroup class.

The second Parameter (restrictions_group) calls the excluded_options method which provides us with the ExclusionList Json Object to actually restrict the properties.

This Method provides us with the actual Json Object which we need to restrict the properties:
exclusion_dict = {"$ENTITY_SET$":{".*": {"Properties":["ProductID"],"Nav_Properties":[]}}}

This is what we require to restrict the properties and not the previous object.

phanak-sap commented on tests/integration/url_generator_only/test_url_generator_only.py:
```diff
@@ -156,38 +168,137 @@ def test_function_imports():
+    random.seed(30)
+    entities , queryable_factory = builder(method_name)
+    methodList = []
+    methodList.clear()
```
is this needed? every parametrized test should be new instance with new clear methodList from line 226

deev-pal08 replied:
This won't be required, will remove this in the next commit..

deev-pal08 committed:
Updated entities.py and test_url_generator_only.py #122

Resolved two issues mentioned in the PR comments:

1. There were redundant restriction object passed as parameters to the $Filter and $Orderby Query Options, which was not required as one class up all the data are provided in the single restrictions parameter.

Issue: https://github.com/SAP/odfuzz/pull/122#discussion_r1068078494

2. Every parametrized test in test_url_generator_only.py should be new instance with new clear methodList and hence we do not need to clear the methodList exclusively.

Issue: https://github.com/SAP/odfuzz/pull/122#discussion_r1068367415

phanak-sap commented on odfuzz/entities.py:
```diff
@@ -888,6 +909,30 @@ def generate(self, depending_data):
+                            proprty_index -= 1
+                        proprty_index += 1
+    
+    def _regex_match(self, entity_name, proprty_name):
```
same.. rename to better self-documenting name

deev-pal08 committed:
Update entities.py #122

Renamed the method self._regex_match() to better self-documenting name which is self._check_for_restricted_properties()

Resolving issue for comment:
 https://github.com/SAP/odfuzz/pull/122#discussion_r1072779200
https://github.com/SAP/odfuzz/pull/122#discussion_r1072779695
